src/Microsoft.Health.Dicom.Core/Features/CustomTag/IndexableDicomTagService.cs

```diff
@@ -0,0 +1,68 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Dicom;
+using EnsureThat;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Dicom.Core.Configs;
+using Microsoft.Health.Dicom.Core.Extensions;
+using Microsoft.Health.Dicom.Core.Features.Query;
+
+namespace Microsoft.Health.Dicom.Core.Features.CustomTag
+{
+    public class IndexableDicomTagService : IIndexableDicomTagService
+    {
+        private readonly ICustomTagStore _customTagStore;
+        private readonly bool _enableCustomQueryTags;
+        public static readonly IReadOnlyList<IndexableDicomTag> CoreIndexableDicomTags = GetCoreIndexableDicomTags();
+        private List<IndexableDicomTag> _allIndexableDicomTags;
+
+        public IndexableDicomTagService(ICustomTagStore customTagStore, IOptions<FeatureConfiguration> featureConfiguration)
+        {
+            EnsureArg.IsNotNull(customTagStore, nameof(customTagStore));
+            EnsureArg.IsNotNull(featureConfiguration?.Value, nameof(featureConfiguration));
+            _customTagStore = customTagStore;
+            _enableCustomQueryTags = featureConfiguration.Value.EnableCustomQueryTags;
+        }
+
+        public async Task<IReadOnlyCollection<IndexableDicomTag>> GetIndexableDicomTagsAsync(CancellationToken cancellationToken = default)
+        {
+            if (_enableCustomQueryTags)
+            {
+                if (_allIndexableDicomTags == null)
```

Comment:
You will have to lock it so, that one request wins and sets this value

Reply:
ex: C:\repos\dicom-server\src\Microsoft.Health.Dicom.Core\Features\Store\Entries\StreamOriginatedDicomInstanceEntry.cs

Reply:
ThreadSafe? I was thinking Scoped instance not likely to hit that problem.